pkg/operator/kit/resource.go

```diff
@@ -22,140 +22,162 @@ package kit
+	}
+	_, err = clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Create(crd)
+	if err != nil {
+		if !errors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create %s CRD. %+v", resource.Name, err)
 		}
 	}
 
+	// wait for CRD being established
+	err = wait.Poll(interval, timeout, func() (bool, error) {
+		crd, err = clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crdName, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		for _, cond := range crd.Status.Conditions {
+			switch cond.Type {
+			case apiextensionsv1beta1.Established:
+				if cond.Status == apiextensionsv1beta1.ConditionTrue {
+					return true, err
+				}
+			case apiextensionsv1beta1.NamesAccepted:
+				if cond.Status == apiextensionsv1beta1.ConditionFalse {
+					fmt.Printf("Name conflict: %v\n", cond.Reason)
+				}
+			}
+		}
+		return false, err
+	})
+	if err != nil {
+		deleteErr := clientset.ApiextensionsV1beta1().CustomResourceDefinitions().Delete(crdName, nil)
+		if deleteErr != nil {
+			return errorsUtil.NewAggregate([]error{err, deleteErr})
+		}
+		return err
+	}
 	return nil
 }
 
-// CreateCustomResource creates a single custom resource, but does not wait for it to initialize
-func CreateCustomResource(context KubeContext, resource CustomResource) error {
-	logger.Infof("creating %s resource", resource.Name)
-	r := &v1beta1.ThirdPartyResource{
+func createTPR(resource CustomResource, clientset kubernetes.Interface) error {
+	logger.Infof("creating %s TPR", resource.Name)
+	tprName := fmt.Sprintf("%s.%s", resource.Name, resource.Group)
+	tpr := &v1beta1.ThirdPartyResource{
 		ObjectMeta: metav1.ObjectMeta{
-			Name: fmt.Sprintf("%s.%s", resource.Name, resource.Group),
+			Name: tprName,
 		},
 		Versions: []v1beta1.APIVersion{
 			{Name: resource.Version},
 		},
-		Description: resource.Description,
+		Description: fmt.Sprintf("ThirdPartyResource for Rook %s", resource.Name),
 	}
-	_, err := context.Clientset.ExtensionsV1beta1().ThirdPartyResources().Create(r)
+	_, err := clientset.ExtensionsV1beta1().ThirdPartyResources().Create(tpr)
 	if err != nil {
 		if !errors.IsAlreadyExists(err) {
-			return fmt.Errorf("failed to create %s resource. %+v", resource.Name, err)
+			return fmt.Errorf("failed to create %s TPR. %+v", resource.Name, err)
 		}
 	}
 
-	return nil
-}
-
-func waitForCustomResourceInit(context KubeContext, resource CustomResource) error {
-	restcli := context.Clientset.CoreV1().RESTClient()
-	uri := resourceURI(resource, "")
-	return Retry(context, func() (bool, error) {
+	// wait for TPR being established
+	restcli := clientset.CoreV1().RESTClient()
+	uri := fmt.Sprintf("apis/%s/%s/%s", resource.Group, resource.Version, resource.Plural)
+	err = wait.Poll(interval, timeout, func() (bool, error) {
 		_, err := restcli.Get().RequestURI(uri).DoRaw()
 		if err != nil {
-			logger.Infof("did not yet find resource %s at %s. %+v", resource.Name, uri, err)
```

Comment:
I didnt find it useful. It was giving a lot of noise. I can set it to logger.debugf. But i think it wasnt useful since the poll function will tell us if it times out or if it succeeds.